pkg/controller/druid/handler.go

```diff
@@ -879,6 +926,26 @@ func namespacedName(name, namespace string) *types.NamespacedName {
 	return &types.NamespacedName{Name: name, Namespace: namespace}
 }
 
+// IntPointer returns a pointer to int32.
+func IntPointer(i int32) *int32 {
+	return &i
+}
+
+// GetHPAReplicaCountOrDefault returns the count of pods in case hpa is mentioned.
+func GetHPAReplicaCountOrDefault(client client.Client, name types.NamespacedName, defaultReplicaCount int32) int32 {
```

Comment:
So, I read a bit more about horizontal pod autoscaler and I think this shouldn't be needed. From what  I  understood, replica count set  on StatefulSet  doesn't matter when an autoscaler  is attached  to that StatefulSet.

Did you really observe operator  continuously updating the  StatefulSet replicas when autoscaler chose a different number of  replicas ?
or what goes wrong   if we  don't  do this?

Reply:
@himanshug 
-> As mentioned in PR, with HPA deployed, hpa controller is responsible for maintaining replica count. The operator does not reconcile it back. 
-> Why we need this function is, if i do any change in CR lets say adjust limits and apply it, the replica count mentioned in the CR spec get deployed.....hpa again brings it back to the count mentioned in hpa spec. 
-> Their is shuffle pods, and recreation takes place. Plus CR should be the only spec file need to be maintained for the whole cluster. And the count should reflect both in CR as well as HPA at all times exactly the same.
-> With removing this function, this can happen. CR is showing 2 count for middlemanager ( as initially we deployed with 2 replica) and with hpa enabled replica is showing 5 ( as minreplica count is 5 ).
Here is why i feel that this function is important.

Reply:
@himanshug a test is failing due this function. Not sure if the issue is in function, it's running fine in cluster. Did you find any specific issue in the function ?

Reply:
I think build is failing because stateful construction now requires a get to k8s  server to 
 succeed which doesn't exist when unit tests run.

I think main question is, is there any significance to `nodeSpec.replicas` attribute , wouldn't k8s just ignore that on the StatefulSet when HPA exists for that StatefulSet ? And, in case, we need to set that to min replica value as in HPA spec, then why goto k8s api server to get it ..rather read it from the hpa spec we already have in the cr ?

Reply:
@himanshug Yes when ```minRelicas``` declared in the spec, k8s would take that as source of truth for the replica count. As per my practice, i don't mention the replica count in the spec when using HPA as its more confusing to have different counts. 
At the same time your point seem LGMT we can take the count from the spec itself. But i still feel as per practice no need to mention replicas when using HPA. 
Your call either works.

Reply:
we definitely shouldn't contact the k8s server to get hpa when we already have it.

on whether to try/adjust same on StatefulSet , I think it is ok to not change anything for StatefulSet, users know that they have configured hpa and that replica count on StatefulSet is ignored by k8s in this case. It would be same experience if user deployed Druid cluster without using this operator.

Reply:
Agreed, i'll remove the client context part and get the value directly from HPA spec.

Reply:
Done changes, now not using client context for getting the default count. Getting directly from ```nodeSpec.Replicas```. I have tested this on my cluster, works for me.